Updated parsing for DX files

tluchko opened the pull request:
While using GridDataFormats, I ran into a couple of issues.

1. The library couldn't read in some of the files it produced.  I tracked this down to lines that look like
```
0.0     -1e-06  0.0     
```
In particular, the parser was splitting `-1e-6` into `-1` and `e-6`.  I modified the `FLOAT` regular expression but this required also modifying parts of the code that expected `INTEGER` as all `int`s  are also `float`s.

2. Parsing large files was pretty slow.  On my system, it would take 45 s to read a 74 MB file. I changed just the loop where the data is read to avoid invoking the parser and doing the type casting.  This got the total read time down to about 7 s.  If you are willing to assume that the data has three entries per line (except the last) you can get this down to about 4 s, of which 2 s is spent reading in the data.  My commit has the slower code that doesn't make this assumption and I've commented out the faster version.

For comparison, just reading the file into a list of strings takes about 1.5 s using the code
```
s.extend([line.split() for line in fh])
```
where `s` is an empty list and `fh` is an open file handle.

I wasn't sure how to run your test suite but I've used these changes on many of my own files and they work fine.

Hopefully you find this useful,

Tyler

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -885,8 +888,20 @@ def __array(self):
                 raise DXParseError("array: missing number of items")
-            self.currentobject['array'] = [self.__consume().value('REAL') \
-                                           for i in range(self.currentobject['size'])]
+            # This is the slow part.  Once we get here, we are just
```
The performance improvement here looks very good.

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -775,10 +773,12 @@ def __gridpositions(self):
             shape = []
             try:
-                while self.__peek().iscode('INTEGER'):
+                while True:
```
using `try/except` instead of `if` looks good to me

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -775,10 +773,12 @@ def __gridpositions(self):
-                    shape.append(tok.value())
-            except DXParserNoTokens:
+                    shape.append(tok.value('INTEGER'))
+            except (DXParserNoTokens,ValueError):
```
PEP8: space after comma

tluchko replied:
Fixed in `64fc3ae`

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -829,10 +828,12 @@ def __gridconnections(self):
-                    shape.append(tok.value())
-            except DXParserNoTokens:
+                    shape.append(tok.value('INTEGER'))
+            except (DXParserNoTokens,ValueError):
```
PEP8: space after comma

tluchko replied:
Fixed in `64fc3ae`

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -885,8 +888,20 @@ def __array(self):
+            # reading in a long list of numbers.  Conversion to floats
+            # will be done later when the numpy array is created.
+
+            # Don't assume anything about whitespace or the number of elements per row
```
Yes – keep it more general. Thanks for the comments in the code.

tluchko replied:
Sorry, do you want me to delete the alternate version or leave it as is?

orbeckst replied:
Leave it.

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -611,10 +611,8 @@ class DXParser(object):
-          (\d+\.\d*([eE][-+]\d+)?)  # scientific notation)
-          |(\d*\.\d+([eE][-+]\d+)?)
-          |(\d[eE][-+]\d+))
-    |(?P<INTEGER>[-+]?\d+)       # integer
```
I am not thrilled throwing out the INTEGER pattern. Could you add a test case with an example that fails with the original code, something that contains
```
0.0     -1e-06  0.0     
```
Maybe just update [gridData/tests/test.dx](https://github.com/MDAnalysis/GridDataFormats/blob/master/gridData/tests/test.dx) with some offending lines or add a new minimal tests file.

Basically, we want to have a test that shows that the bug has been squashed.

orbeckst replied:
It should be possible to fix the regular expressions... I probably made a subtle mistake when I wrote them a long time ago.

tluchko replied:
I modified `test.dx` in `0fabaa9`and `3598ae5`.  I don't know what framework you are using for your test suite or how to run the tests.  I am really only familiar with the `unittest` module. I can confirm that this test file now loads correctly but would fail at multiple points in the previous version.

I tried to keep the `INTEGER` pattern but anything that is an integer is captured by `FLOAT` since regular expressions are greedy.  I only deleted it because it didn't do anything after I updated the `FLOAT` expression. It might be possible to get `INTEGER` to work by looking ahead or back but I didn't investigate this. BTW, I just took the `FLOAT` regex from the [`re`](https://docs.python.org/3/library/re.html#simulating-scanf) documentation.

orbeckst replied:
Ok, thanks for the explanation. Then let's go with your modifications and leave out INTEGER.

Can you add a comment to the regexes that references the source as the re module with the link that you gave and that states that FLOAT will also catch integers?

tluchko replied:
Done.

tluchko commented:
I also updated AUTHOR and CHANGELOG.  Following previous entries, I added new version number in CHANGELOG but I did not try to update this anywhere else.

orbeckst commented:
@tluchko the test [gridData.tests.test_dx.test_read_dx](https://travis-ci.org/MDAnalysis/GridDataFormats/jobs/414726557#L838) fails. Do you know why? Does it indicate that the bug is not fixed – or do you have to update the test so that it reflects the expected result?

orbeckst commented:
In the tests at https://github.com/MDAnalysis/GridDataFormats/blob/030628dfd458bfc4e1936f4e807a3e6272994e44/gridData/tests/test_dx.py#L15 you need to replace the `np.ones(POINTS)` with something like
```python
ref = np.ones(POINTS)
ref[4] = 1e-6
ref[5] =  1e+6
assert_equal(g.grid.flat, ref)
```

tluchko commented:
Is there a way to run the test suite locally?  Trying to guess and test by pushing is not working for me.

orbeckst commented:
P.S.: Only the Travis tests have to pass, the AppVeyor Windows builds are not properly configured at the moment (see #47).

orbeckst commented:
@tluchko thank you for your first contribution!
